src/coprocessor/dag/batch/executors/selection_executor.rs

```diff
@@ -0,0 +1,639 @@
+        self.summary_collector
+            .on_finish_batch(timer, result.data.rows_len());
+        result
+    }
+
+    #[inline]
+    fn collect_statistics(&mut self, destination: &mut BatchExecuteStatistics) {
+        self.src.collect_statistics(destination);
+        self.summary_collector
+            .collect_into(&mut destination.summary_per_executor);
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use cop_datatype::FieldTypeTp;
+
+    use crate::coprocessor::codec::batch::LazyBatchColumnVec;
+    use crate::coprocessor::codec::data_type::VectorValue;
+    use crate::coprocessor::dag::batch::executors::util::mock_executor::MockExecutor;
+    use crate::coprocessor::dag::expr::EvalWarnings;
+    use crate::coprocessor::dag::rpn_expr::types::RpnFnCallPayload;
+    use crate::coprocessor::dag::rpn_expr::RpnFunction;
+
+    #[test]
+    fn test_empty_rows() {
+        #[derive(Debug, Clone, Copy)]
+        struct FnFoo;
+
+        impl RpnFunction for FnFoo {
+            fn name(&self) -> &'static str {
+                "FnFoo"
+            }
+
+            fn args_len(&self) -> usize {
+                0
+            }
+
+            fn eval(
+                &self,
+                _rows: usize,
+                _context: &mut EvalContext,
+                _payload: RpnFnCallPayload<'_>,
+            ) -> Result<VectorValue> {
+                // This function should never be called because we filter no rows
+                unreachable!()
+            }
+
+            fn box_clone(&self) -> Box<dyn RpnFunction> {
+                Box::new(*self)
+            }
+        }
+
+        let src_exec = MockExecutor::new(
+            vec![],
+            vec![
+                BatchExecuteResult {
+                    data: LazyBatchColumnVec::empty(),
+                    warnings: EvalWarnings::default(),
+                    is_drained: Ok(false),
+                },
+                BatchExecuteResult {
+                    data: LazyBatchColumnVec::empty(),
+                    warnings: EvalWarnings::default(),
+                    is_drained: Ok(true),
+                },
+            ],
+        );
+
+        let mut exec = BatchSelectionExecutor::new_for_test(
+            src_exec,
+            vec![RpnExpressionBuilder::new()
+                .push_fn_call(FnFoo, FieldTypeTp::LongLong)
+                .build()],
+        );
+
+        // When source executor returns empty rows, selection executor should process correctly.
+        // No errors should be generated and the predicate function should not be called.
+
+        let r = exec.next_batch(1);
+        // The scan rows parameter has no effect for mock executor. We don't care.
+        assert_eq!(r.data.rows_len(), 0);
+        assert!(!r.is_drained.unwrap());
+
+        let r = exec.next_batch(1);
+        assert_eq!(r.data.rows_len(), 0);
+        assert!(r.is_drained.unwrap());
+    }
+
+    fn make_src_executor_using_fixture_1() -> MockExecutor {
```

Comment:
Could you add some comments to describe it, so we needn't always read the code when we want to know what it will returns, maybe like
```
     * The first batch:
     *      1    None
     *      None 7.0
     * The second batch:
     *      empty
     * The last batch
     *      None, None
```